File: hash/hash.py
# -*- coding: utf-8 -*-
"""
Created on Wed Apr 17 14:35:14 2019

@author: Administrator
"""

class hash_table:
    def __init__(self):
        self.table = [None]*127
        
        
    #hash function
    def Hash_func(self, value):
        key = 0
        for i in range(0, len(value)):
            key += ord(value[i])
        return key % 127
    
    def Insert(self, value):
        hash = self.Hash_func(value)
        if self.table[hash] is None:
            self.table[hash] = value
            
    def Search(self,value):
        hash = self.Hash_func(value)
        if self.table[hash] is None:
            return None
        else:
            return hex(id(self.table[hash]))
    
    def Remove(self,value):
        hash = self.Hash_func(value)
        if self.table[hash] is None:
            print("没有找到 value",value)
        else:
            print("这个 value",value,"被删除")
            self.table[hash] is None
            


H=hash_table()
H.Insert("A")
H.Insert("B")
H.Insert("C")

print(H.Search("C"))

File: hash/hash.cpp
#include <iostream>
#include <string>
using namespace std;

const int TABLE_SIZE = 127;

class HashEntry
{
public:
	string value;
	HashEntry(string value)
	{
		this->value = value;
	}
};


class HashMap
{
private:
	HashEntry **table;
public:
	HashMap()
	{
		table = new HashEntry*[TABLE_SIZE];
		for(int k=0; k < TABLE_SIZE; k++)
		{
			table[k] = NULL;
		}
	}

	//hash function
	int HashFunc(string value)
	{
		int key = 0;
		for(int i=0; i<value.size();i++)
		{
			key += value[i];
		}
		return key % TABLE_SIZE;
	}

	void Insert(string value)
	{
		int hash = HashFunc(value);
		if(table[hash] == NULL)
		{
			table[hash] = new HashEntry(value);
		}
	}

	HashEntry* Search(string value)
	{
		int hash = HashFunc(value);
		if(table[hash] == NULL)
			return NULL;
		else
			return table[hash];
	}

	void Remove(string value)
	{
		int hash = HashFunc(value);
		if(table[hash] == NULL)
		{
			cout<<"没有发现元素"<<value<<endl;
		}
		else
		{
			table[hash] == NULL;
			cout<<"元素值"<<value<<"被删除"<<endl;
		}

	}


};


int main()
{
	HashMap hash;
	hash.Insert("one");
	hash.Insert("two");
	hash.Insert("three");

	cout<<hash.Search("two");
	return 0;
}

